File: apps/spindrift-verifier/pkg/verifier/verify.go
package verifier

import (
	"crypto/ed25519"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
	"time"
)

var slsaVersionRegex = regexp.MustCompile(`(?i)slsa[^/]*/v(\d+(?:\.\d+)?)`)

// Verify performs strict provenance verification against specified expectations.
func Verify(req VerificationRequest, now func() time.Time) VerificationResponse {
	if now == nil {
		now = time.Now
	}

	backend := req.Expectations.Backend
	if backend == "" {
		backend = "hosted"
	}

	// 1. Check for missing provenance
	if len(req.Provenance.Statement) == 0 || string(req.Provenance.Statement) == "null" {
		return VerificationResponse{
			Version: "v1",
			OK:      false,
			Code:    "PROVENANCE_MISSING",
			Message: fmt.Sprintf("%s returned no backend provenance", backend),
		}
	}

	// 2. Check artifact digest & reference
	if req.Artifact.Digest == "" {
		return VerificationResponse{
			Version: "v1",
			OK:      false,
			Code:    "PROVENANCE_INVALID",
			Message: fmt.Sprintf("%s returned no immutable image reference for digest", backend),
		}
	}

	hasImmutableRef := false
	for _, ref := range req.Artifact.Refs {
		if strings.Contains(ref, req.Artifact.Digest) {
			hasImmutableRef = true
			break
		}
	}
	if len(req.Artifact.Refs) > 0 && !hasImmutableRef {
		return VerificationResponse{
			Version: "v1",
			OK:      false,
			Code:    "PROVENANCE_INVALID",
			Message: fmt.Sprintf("%s returned no immutable image reference for %s", backend, req.Artifact.Digest),
		}
	}

	// 3. Parse statement JSON
	var stmt map[string]interface{}
	if err := json.Unmarshal(req.Provenance.Statement, &stmt); err != nil {
		return VerificationResponse{
			Version: "v1",
			OK:      false,
			Code:    "PROVENANCE_INVALID",
			Message: fmt.Sprintf("%s provenance did not verify: invalid JSON statement", backend),
		}
	}

	// 4. Extract builder ID
	extractedBuilderID := extractBuilderID(stmt)
	if extractedBuilderID != "" && req.Expectations.ExpectedBuilderID != "" && extractedBuilderID != req.Expectations.ExpectedBuilderID {
		return VerificationResponse{
			Version: "v1",
			OK:      false,
			Code:    "PROVENANCE_INVALID",
			Message: fmt.Sprintf("%s provenance builder mismatch: expected %s, got %s", backend, req.Expectations.ExpectedBuilderID, extractedBuilderID),
		}
	}

	// 4b. Provenance authenticity: when the route configured a builder public
	// key, the backend's signature over the exact statement bytes must verify
	// against it. This is the cryptographic half of "provenance
	// authenticity" — a statement nobody signed, or signed by a different key,
	// or modified after signing, is rejected with no silent fallback.
	if req.Expectations.BuilderPublicKey != "" {
		var sig []byte
		if req.Provenance.Signature != "" {
			decoded, err := base64.StdEncoding.DecodeString(req.Provenance.Signature)
			if err != nil {
				return VerificationResponse{
					Version: "v1",
					OK:      false,
					Code:    "PROVENANCE_INVALID",
					Message: fmt.Sprintf("%s provenance signature is not base64: %v", backend, err),
				}
			}
			sig = decoded
		}
		if len(sig) == 0 {
			return VerificationResponse{
				Version: "v1",
				OK:      false,
				Code:    "PROVENANCE_INVALID",
				Message: fmt.Sprintf("%s provenance carries no signature for a route that requires authenticity", backend),
			}
		}
		pubDER, err := base64.StdEncoding.DecodeString(req.Expectations.BuilderPublicKey)
		if err != nil {
			return VerificationResponse{
				Version: "v1",
				OK:      false,
				Code:    "PROVENANCE_INVALID",
				Message: fmt.Sprintf("configured builder public key is not base64: %v", err),
			}
		}
		pubIface, err := x509.ParsePKIXPublicKey(pubDER)
		if err != nil {
			return VerificationResponse{
				Version: "v1",
				OK:      false,
				Code:    "PROVENANCE_INVALID",
				Message: fmt.Sprintf("configured builder public key is not parseable: %v", err),
			}
		}
		pub, ok := pubIface.(ed25519.PublicKey)
		if !ok {
			return VerificationResponse{
				Version: "v1",
				OK:      false,
				Code:    "PROVENANCE_INVALID",
				Message: "configured builder public key is not an Ed25519 key",
			}
		}
		if !ed25519.Verify(pub, req.Provenance.Statement, sig) {
			return VerificationResponse{
				Version: "v1",
				OK:      false,
				Code:    "PROVENANCE_INVALID",
				Message: fmt.Sprintf("%s provenance signature does not verify against the statement (tampered or signed by another key)", backend),
			}
		}
	}

	// 4c. Subject binding: when the statement names a subject, its digest must
	// be the artifact being admitted. A provenance document that names a
	// different artifact does not bind this one, and "wrong subject" is exactly
	// the failure that catches.
	if subjectDigest := extractSubjectDigest(stmt); subjectDigest != "" {
		if subjectDigest != req.Artifact.Digest {
			return VerificationResponse{
				Version: "v1",
				OK:      false,
				Code:    "PROVENANCE_INVALID",
				Message: fmt.Sprintf("%s provenance subject names %q, not the admitted artifact %s", backend, subjectDigest, req.Artifact.Digest),
			}
		}
	}

	// 5. Extract bundle digest and verify binding
	extractedBundleDigest := extractBundleDigest(stmt)
	if req.Expectations.BundleDigest != "" {
		if extractedBundleDigest == "" {
			return VerificationResponse{
				Version: "v1",
				OK:      false,
				Code:    "PROVENANCE_INVALID",
				Message: fmt.Sprintf("%s verified provenance does not bind the source bundle digest", backend),
			}
		}
		if extractedBundleDigest != req.Expectations.BundleDigest {
			return VerificationResponse{
				Version: "v1",
				OK:      false,
				Code:    "PROVENANCE_INVALID",
				Message: fmt.Sprintf("%s verified provenance names bundle %s, not %s", backend, extractedBundleDigest, req.Expectations.BundleDigest),
			}
		}
	}

	// 6. Calculate achieved build level and compare with policy
	achievedLevel := req.Provenance.ClaimedLevel
	if req.Expectations.MaximumLevel > 0 && achievedLevel > req.Expectations.MaximumLevel {
		achievedLevel = req.Expectations.MaximumLevel
	}

	if req.Expectations.MinimumLevel > 0 && achievedLevel < req.Expectations.MinimumLevel {
		return VerificationResponse{
			Version: "v1",
			OK:      false,
			Code:    "BUILD_LEVEL_BELOW_POLICY",
			Message: fmt.Sprintf("%s produced verified Build Level %d, but this Target requires L%d", backend, achievedLevel, req.Expectations.MinimumLevel),
		}
	}

	// 7. Extract SLSA version
	slsaVer := extractSLSAVersion(stmt)

	builderID := extractedBuilderID
	if builderID == "" {
		builderID = req.Expectations.ExpectedBuilderID
	}

	return VerificationResponse{
		Version: "v1",
		OK:      true,
		Assessment: &BackendProvenanceAssessment{
			ArtifactDigest: req.Artifact.Digest,
			BundleDigest:   extractedBundleDigest,
			Backend:        backend,
			BuilderID:      builderID,
			SLSAVersion:    slsaVer,
			AchievedLevel:  achievedLevel,
			VerifiedAt:     now().UTC().Format(time.RFC3339Nano),
			Envelope:       req.Provenance.Statement,
		},
	}
}

func extractBuilderID(stmt map[string]interface{}) string {
	predicate, _ := stmt["predicate"].(map[string]interface{})
	if predicate == nil {
		return ""
	}
	runDetails, _ := predicate["runDetails"].(map[string]interface{})
	if runDetails == nil {
		return ""
	}
	builder, _ := runDetails["builder"].(map[string]interface{})
	if builder == nil {
		return ""
	}
	id, _ := builder["id"].(string)
	return id
}

func extractBundleDigest(stmt map[string]interface{}) string {
	predicate, _ := stmt["predicate"].(map[string]interface{})
	if predicate == nil {
		return ""
	}
	buildDefinition, _ := predicate["buildDefinition"].(map[string]interface{})
	if buildDefinition == nil {
		return ""
	}
	externalParameters, _ := buildDefinition["externalParameters"].(map[string]interface{})
	if externalParameters == nil {
		return ""
	}
	digest, _ := externalParameters["bundleDigest"].(string)
	return digest
}

func extractSLSAVersion(stmt map[string]interface{}) string {
	predType, _ := stmt["predicateType"].(string)
	if predType == "" {
		return "unknown"
	}
	match := slsaVersionRegex.FindStringSubmatch(predType)
	if len(match) > 1 {
		return match[1]
	}
	return "1.2"
}

// extractSubjectDigest returns the sha256 digest the statement names as its
// subject, as a "sha256:<hex>" string, or "" when the statement names none.
// The subject binds the provenance to the artifact it claims to describe.
func extractSubjectDigest(stmt map[string]interface{}) string {
	subjects, _ := stmt["subject"].([]interface{})
	if len(subjects) == 0 {
		return ""
	}
	first, _ := subjects[0].(map[string]interface{})
	if first == nil {
		return ""
	}
	digest, _ := first["digest"].(map[string]interface{})
	if digest == nil {
		return ""
	}
	h, _ := digest["sha256"].(string)
	if h == "" {
		return ""
	}
	if strings.HasPrefix(h, "sha256:") {
		return h
	}
	return "sha256:" + h
}

File: apps/spindrift/src/supply-chain/verify.ts
/**
 * Verification of a build backend's SLSA provenance (§16).
 *
 * The build adapter reports evidence; it does not assess itself. This module is
 * the process boundary around the pinned `slsa-verifier` binary and returns the
 * normalized facts core may persist after that binary accepts the envelope.
 */
import { mkdtemp, rm, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import type {
  BuildLevel,
  BuildProvenance,
  BuildSource,
} from '../adapters/build/contract.ts';
import type { Artifact } from '../domain/desired-state.ts';

export interface ProcessResult {
  readonly exitCode: number;
  readonly stdout: string;
  readonly stderr: string;
}

/** The native process seam; tests replace the far side, never this module. */
export interface ProcessExecutor {
  run(command: readonly string[]): Promise<ProcessResult>;
}

/**
 * The reference that pins an artifact to its digest, whatever it is made of.
 *
 * `immutableImageRef` is the deploy side's version of this and gates on
 * `type === 'image'`, which is right there — a workload needs an image. It is
 * wrong here. §16 says core "signs that digest", and the digest of a `files`
 * artifact is as signable as an image's; the property this module needs is only
 * that the reference names the digest rather than a tag, which is what closes
 * the check/use race the verifier's own contract warns about. Gating on image
 * meant every static-site build was refused before the verifier was spawned.
 */
export function digestPinnedRef(artifact: Artifact): string | null {
  return (
    artifact.refs.find((ref) => ref.endsWith(`@${artifact.digest}`)) ?? null
  );
}

/** Execute one pinned tool without a shell. */
export const bunProcessExecutor: ProcessExecutor = {
  async run(command) {
    const child = Bun.spawn([...command], {
      stdout: 'pipe',
      stderr: 'pipe',
      env: Bun.env,
    });
    const [stdout, stderr, exitCode] = await Promise.all([
      new Response(child.stdout).text(),
      new Response(child.stderr).text(),
      child.exited,
    ]);
    return { exitCode, stdout, stderr };
  },
};

/** Facts copied out of evidence core successfully assessed. */
export interface BackendProvenanceAssessment {
  readonly artifactDigest: string;
  readonly bundleDigest: string;
  readonly backend: string;
  readonly builderId: string;
  readonly slsaVersion: string;
  readonly achievedLevel: BuildLevel;
  readonly verifiedAt: string;
  /** The durable envelope, not only the normalized summary. */
  readonly envelope: unknown;
}

export interface VerifyProvenanceInput {
  readonly artifact: Artifact;
  readonly provenance: BuildProvenance;
  readonly backend: string;
  /** Trusted expectation from the route profile, never from the envelope. */
  readonly expectedBuilderId: string;
  /** The strongest claim this code-defined route profile permits. */
  readonly maximumLevel: BuildLevel;
  readonly source: BuildSource;
}

export type ProvenanceRefusalCode = 'PROVENANCE_MISSING' | 'PROVENANCE_INVALID';

export type ProvenanceVerification =
  | { readonly ok: true; readonly assessment: BackendProvenanceAssessment }
  | {
      readonly ok: false;
      readonly code: ProvenanceRefusalCode;
      readonly message: string;
    };

export interface ProvenanceVerifier {
  verify(input: VerifyProvenanceInput): Promise<ProvenanceVerification>;
}

export interface SlsaVerifierOptions {
  readonly executable?: string;
  readonly processes?: ProcessExecutor;
  readonly now?: () => Date;
}

/**
 * The production verifier.
 *
 * It verifies an immutable reference and a copied envelope. No tag reaches the
 * tool, closing the check/use race the verifier's own contract warns about.
 *
 * **Two limitations of the `verify-image` path, recorded rather than hidden.**
 * `apps/spindrift-verifier/main.go` builds that path's request with
 * `ClaimedLevel: 2`, `MinimumLevel: 1`, `MaximumLevel: 2` and
 * `Backend: "hosted"` hardcoded, and nothing on the command line conveys any of
 * them:
 *
 * 1. *The level is decided here, not there.* Those constants make the binary's
 *    own level arithmetic inert on this path — `min(2, 2) = 2 >= 1` always
 *    passes — so the achieved level is entirely {@link lowerLevel} below,
 *    against the route profile's ceiling. That matches §16 ("guarantees belong
 *    to code-defined backend/runner profiles"), but it does mean an L3 claim
 *    rests on {@link CloudBuildRoute}'s constant and on nothing the envelope
 *    says. Passing the level through would need a flag on both sides and a new
 *    verifier image; it would not change any verdict today.
 * 2. *`--source-uri` is accepted and ignored.* `Expectations.SourceURI` is
 *    declared in `pkg/verifier/types.go` and never read by `Verify`, so the
 *    only source binding anything checks is {@link bundleDigestOf} below.
 */
export class SlsaVerifier implements ProvenanceVerifier {
  private readonly executable: string;
  private readonly processes: ProcessExecutor;
  private readonly now: () => Date;

  constructor(options: SlsaVerifierOptions = {}) {
    this.executable = options.executable ?? 'spindrift-verifier';
    this.processes = options.processes ?? bunProcessExecutor;
    this.now = options.now ?? (() => new Date());
  }

  async verify(input: VerifyProvenanceInput): Promise<ProvenanceVerification> {
    if (input.provenance.statement === null) {
      return {
        ok: false,
        code: 'PROVENANCE_MISSING',
        message: `${input.backend} returned no backend provenance`,
      };
    }
    const immutableRef = digestPinnedRef(input.artifact);
    if (immutableRef === null) {
      return {
        ok: false,
        code: 'PROVENANCE_INVALID',
        message: `${input.backend} returned no immutable reference for ${input.artifact.digest}`,
      };
    }

    const directory = await mkdtemp(join(tmpdir(), 'spindrift-provenance-'));
    const provenancePath = join(directory, 'provenance.json');
    try {
      await writeFile(
        provenancePath,
        JSON.stringify(input.provenance.statement),
        { mode: 0o600 },
      );
      const command = [
        this.executable,
        'verify-image',
        immutableRef,
        '--provenance-path',
        provenancePath,
        '--source-uri',
        sourceUriOf(input.source),
        '--print-provenance',
        '--builder-id',
        input.expectedBuilderId,
      ];
      const result = await this.processes.run(command);
      if (result.exitCode !== 0) {
        return {
          ok: false,
          code: 'PROVENANCE_INVALID',
          message: verifierFailure(input.backend, result),
        };
      }

      const envelope = verifiedEnvelope(result.stdout);
      if (envelope === null) {
        return {
          ok: false,
          code: 'PROVENANCE_INVALID',
          message: `${input.backend} verifier returned no readable provenance`,
        };
      }
      const bundleDigest = bundleDigestOf(envelope);
      if (bundleDigest !== input.source.bundleDigest) {
        return {
          ok: false,
          code: 'PROVENANCE_INVALID',
          message:
            bundleDigest === null
              ? `${input.backend} verified provenance does not bind the source bundle digest`
              : `${input.backend} verified provenance names bundle ${bundleDigest}, not ${input.source.bundleDigest}`,
        };
      }

      return {
        ok: true,
        assessment: {
          artifactDigest: input.artifact.digest,
          bundleDigest,
          backend: input.backend,
          builderId: builderIdOf(envelope) ?? input.expectedBuilderId,
          slsaVersion: slsaVersionOf(envelope),
          achievedLevel: lowerLevel(
            input.provenance.claimedLevel,
            input.maximumLevel,
          ),
          verifiedAt: this.now().toISOString(),
          envelope,
        },
      };
    } finally {
      await rm(directory, { recursive: true, force: true });
    }
  }
}

function verifiedEnvelope(stdout: string): unknown | null {
  const output = stdout.trim();
  if (output === '') return null;
  try {
    return JSON.parse(output);
  } catch {
    return null;
  }
}

function lowerLevel(left: BuildLevel, right: BuildLevel): BuildLevel {
  return Math.min(left, right) as BuildLevel;
}

function sourceUriOf(source: BuildSource): string {
  if (source.origin.type === 'archive') return source.origin.location;
  return source.origin.repository
    .replace(/^git\+/, '')
    .replace(/^https?:\/\//, '')
    .replace(/\.git$/, '');
}

function record(value: unknown): Record<string, unknown> | null {
  return typeof value === 'object' && value !== null
    ? (value as Record<string, unknown>)
    : null;
}

function builderIdOf(statement: unknown): string | null {
  const root = record(statement);
  const predicate = record(root?.predicate);
  const runDetails = record(predicate?.runDetails);
  const builder = record(runDetails?.builder);
  const id = builder?.id;
  return typeof id === 'string' && id !== '' ? id : null;
}

function bundleDigestOf(statement: unknown): string | null {
  const root = record(statement);
  const predicate = record(root?.predicate);
  const buildDefinition = record(predicate?.buildDefinition);
  const externalParameters = record(buildDefinition?.externalParameters);
  const digest = externalParameters?.bundleDigest;
  return typeof digest === 'string' && digest !== '' ? digest : null;
}

function slsaVersionOf(statement: unknown): string {
  const predicateType = record(statement)?.predicateType;
  if (typeof predicateType !== 'string') return 'unknown';
  const match = /slsa[^/]*\/v(\d+(?:\.\d+)?)/i.exec(predicateType);
  return match?.[1] ?? 'unknown';
}

function verifierFailure(backend: string, result: ProcessResult): string {
  const detail = result.stderr.trim() || result.stdout.trim();
  return `${backend} provenance did not verify${detail === '' ? '' : `: ${detail}`}`;
}
